[runtimes] [openmp] `-nostdlib++` added to `CMAKE_REQUIRED_FLAGS` breaks hwloc detection for OMPT

To reproduce:

```
$ cmake -G Ninja -DLLVM_ENABLE_RUNTIMES=openmp -DLIBOMP_USE_HWLOC=ON ${path_to_llvm_project}/runtimes
[…]
-- Looking for hwloc.h
-- Looking for hwloc.h - not found
-- Looking for hwloc_topology_init in /usr/lib64/libhwloc.so
-- Looking for hwloc_topology_init in /usr/lib64/libhwloc.so - not found
-- LIBOMP: Could not find hwloc
CMake Error at /home/mgorny/git/llvm-project/openmp/runtime/cmake/LibompUtils.cmake:26 (message):
  LIBOMP: Hwloc requested but not available
Call Stack (most recent call first):
  /home/mgorny/git/llvm-project/openmp/runtime/CMakeLists.txt:387 (libomp_error_say)


-- Configuring incomplete, errors occurred!
```

`CMakeFiles/CMakeConfigureLog.yaml` hints at the problem:

```yaml
  -
    kind: "try_compile-v1"
    backtrace:
      - "/usr/share/cmake/Modules/CheckIncludeFile.cmake:90 (try_compile)"
      - "/home/mgorny/git/llvm-project/openmp/runtime/cmake/config-ix.cmake:351 (check_include_file)"
      - "/home/mgorny/git/llvm-project/openmp/runtime/CMakeLists.txt:276 (include)"
    checks:
      - "Looking for hwloc.h"
    directories:
      source: "/home/mgorny/git/llvm-project/build/CMakeFiles/CMakeScratch/TryCompile-EEYwvY"
      binary: "/home/mgorny/git/llvm-project/build/CMakeFiles/CMakeScratch/TryCompile-EEYwvY"
    cmakeVariables:
      CMAKE_C_FLAGS: " -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -Wno-comment -fdiagnostics-color -Wall -Wcast-qual -Wimplicit-fallthrough -Wsign-compare -Wno-extra -Wno-pedantic -Wno-maybe-uninitialized -fno-semantic-interposition -fdata-sections"
      CMAKE_C_FLAGS_DEBUG: "-g"
      CMAKE_EXE_LINKER_FLAGS: ""
      CMAKE_MODULE_PATH: "/home/mgorny/git/llvm-project/openmp/runtime/cmake;/home/mgorny/git/llvm-project/openmp/cmake;/home/mgorny/git/llvm-project/openmp/../cmake/Modules;/home/mgorny/git/llvm-project/runtimes/cmake;/home/mgorny/git/llvm-project/runtimes/cmake/modules;/home/mgorny/git/llvm-project/runtimes/../cmake;/home/mgorny/git/llvm-project/runtimes/../cmake/Modules;/home/mgorny/git/llvm-project/runtimes/../llvm/cmake;/home/mgorny/git/llvm-project/runtimes/../llvm/cmake/modules"
    buildResult:
      variable: "LIBOMP_HAVE_HWLOC_H"
      cached: true
      stdout: |
        Change Dir: '/home/mgorny/git/llvm-project/build/CMakeFiles/CMakeScratch/TryCompile-EEYwvY'
        
        Run Build Command(s): /usr/bin/ninja -v cmTC_5eb96
        [1/2] /usr/bin/cc   -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -Wno-comment -fdiagnostics-color -Wall -Wcast-qual -Wimplicit-fallthrough -Wsign-compare -Wno-extra -Wno-pedantic -Wno-maybe-uninitialized -fno-semantic-interposition -fdata-sections  -nostdlib++ -nostdinc++ -o CMakeFiles/cmTC_5eb96.dir/CheckIncludeFile.c.o -c /home/mgorny/git/llvm-project/build/CMakeFiles/CMakeScratch/TryCompile-EEYwvY/CheckIncludeFile.c
        FAILED: CMakeFiles/cmTC_5eb96.dir/CheckIncludeFile.c.o 
        /usr/bin/cc   -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -Wno-comment -fdiagnostics-color -Wall -Wcast-qual -Wimplicit-fallthrough -Wsign-compare -Wno-extra -Wno-pedantic -Wno-maybe-uninitialized -fno-semantic-interposition -fdata-sections  -nostdlib++ -nostdinc++ -o CMakeFiles/cmTC_5eb96.dir/CheckIncludeFile.c.o -c /home/mgorny/git/llvm-project/build/CMakeFiles/CMakeScratch/TryCompile-EEYwvY/CheckIncludeFile.c
        cc: error: unrecognized command-line option ‘-nostdlib++’
        ninja: build stopped: subcommand failed.
        
      exitCode: 1
```

i.e. the `nostdlib++` flags gets appended to *both* C and C++ compiler flags, effectively breaking all compilations done in C mode.

CC @mstorsjo as the one who added the logic

I am seeing this when trying to compile llvm 17.0.6. The error is slightly different in that the flag is accepted, however, there are missing symbols. 

```
    kind: "try_compile-v1"
    backtrace:
      - "/gpfs/tp/gwf/gwf_cmt/cmarsh/spack/install/linux-rhel9-skylake_avx512/gcc-12.3.1/cmake-3.29.6-q6iwwch2crujenzcug42iuvx32hi2mz5/share/cmake-3.29/Modules/CheckLibraryExists.cmake:69 (try_compile)"
      - "/gpfs/tp/gwf/gwf_cmt/cmarsh/spack/stage/spack-stage-llvm-17.0.6-siym653rezmksycrvlez7jah5tk72os3/spack-src/openmp/runtime/cmake/config-ix.cmake:349 (check_library_exists)"
      - "/gpfs/tp/gwf/gwf_cmt/cmarsh/spack/stage/spack-stage-llvm-17.0.6-siym653rezmksycrvlez7jah5tk72os3/spack-src/openmp/runtime/CMakeLists.txt:245 (include)"
    checks:
      - "Looking for hwloc_topology_init in /globalhome/cbm038/HPC/project/spack/install/linux-rhel9-skylake_avx512/gcc-12.3.1/hwloc-2.9.3-rpi4timtsf6t32tbrgvhkxcwfqxnywhv/lib/libhwloc.so"
    directories:
      source: "/gpfs/tp/gwf/gwf_cmt/cmarsh/spack/stage/spack-stage-llvm-17.0.6-siym653rezmksycrvlez7jah5tk72os3/spack-build-siym653/runtimes/runtimes-bins/CMakeFiles/CMakeScratch/TryCompile-XydOOq"
      binary: "/gpfs/tp/gwf/gwf_cmt/cmarsh/spack/stage/spack-stage-llvm-17.0.6-siym653rezmksycrvlez7jah5tk72os3/spack-build-siym653/runtimes/runtimes-bins/CMakeFiles/CMakeScratch/TryCompile-XydOOq"
    cmakeVariables:
      CMAKE_CXX_COMPILER_CLANG_SCAN_DEPS: "CMAKE_CXX_COMPILER_CLANG_SCAN_DEPS-NOTFOUND"
      CMAKE_CXX_COMPILER_TARGET: "x86_64-unknown-linux-gnu"
      CMAKE_C_COMPILER_TARGET: "x86_64-unknown-linux-gnu"
      CMAKE_C_FLAGS: "--gcc-install-dir=/cvmfs/soft.computecanada.ca/gentoo/2023/x86-64-v3/usr/lib/gcc/x86_64-pc-linux-gnu/12 -fPIC -fno-semantic-interposition -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -Wimplicit-fallthrough -Wcovered-switch-default -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -Wall -Wcast-qual -Wformat-pedantic -Wimplicit-fallthrough -Wsign-compare -Wno-enum-constexpr-conversion -Wno-extra -Wno-pedantic"
      CMAKE_C_FLAGS_DEBUG: "-g"
      CMAKE_EXE_LINKER_FLAGS: ""
      CMAKE_MODULE_PATH: "/gpfs/tp/gwf/gwf_cmt/cmarsh/spack/stage/spack-stage-llvm-17.0.6-siym653rezmksycrvlez7jah5tk72os3/spack-src/openmp/runtime/cmake;/gpfs/tp/gwf/gwf_cmt/cmarsh/spack/stage/spack-stage-llvm-17.0.6-siym653rezmksycrvlez7jah5tk72os3/spack-src/openmp/cmake;/gpfs/tp/gwf/gwf_cmt/cmarsh/spack/stage/spack-stage-llvm-17.0.6-siym653rezmksycrvlez7jah5tk72os3/spack-src/openmp/../cmake/Modules;/gpfs/tp/gwf/gwf_cmt/cmarsh/spack/stage/spack-stage-llvm-17.0.6-siym653rezmksycrvlez7jah5tk72os3/spack-src/runtimes/cmake;/gpfs/tp/gwf/gwf_cmt/cmarsh/spack/stage/spack-stage-llvm-17.0.6-siym653rezmksycrvlez7jah5tk72os3/spack-src/runtimes/cmake/modules;/gpfs/tp/gwf/gwf_cmt/cmarsh/spack/stage/spack-stage-llvm-17.0.6-siym653rezmksycrvlez7jah5tk72os3/spack-src/runtimes/../cmake;/gpfs/tp/gwf/gwf_cmt/cmarsh/spack/stage/spack-stage-llvm-17.0.6-siym653rezmksycrvlez7jah5tk72os3/spack-src/runtimes/../cmake/Modules;/gpfs/tp/gwf/gwf_cmt/cmarsh/spack/stage/spack-stage-llvm-17.0.6-siym653rezmksycrvlez7jah5tk72os3/spack-src/runtimes/../llvm/cmake;/gpfs/tp/gwf/gwf_cmt/cmarsh/spack/stage/spack-stage-llvm-17.0.6-siym653rezmksycrvlez7jah5tk72os3/spack-src/runtimes/../llvm/cmake/modules;/gpfs/tp/gwf/gwf_cmt/cmarsh/spack/stage/spack-stage-llvm-17.0.6-siym653rezmksycrvlez7jah5tk72os3/spack-build-siym653/lib/cmake/llvm;/gpfs/tp/gwf/gwf_cmt/cmarsh/spack/stage/spack-stage-llvm-17.0.6-siym653rezmksycrvlez7jah5tk72os3/spack-build-siym653/lib/cmake/llvm"
    buildResult:
      variable: "LIBOMP_HAVE_LIBHWLOC"
      cached: true
      stdout: |
        Change Dir: '/gpfs/tp/gwf/gwf_cmt/cmarsh/spack/stage/spack-stage-llvm-17.0.6-siym653rezmksycrvlez7jah5tk72os3/spack-build-siym653/runtimes/runtimes-bins/CMakeFiles/CMakeScratch/TryCompile-XydOOq'
        
        Run Build Command(s): /globalhome/cbm038/HPC/project/spack/install/linux-rhel9-skylake_avx512/gcc-12.3.1/ninja-1.12.0-x2csjlvyad6s7gnfoa5ubpncbt3cu3dg/bin/ninja-build -v cmTC_48e41
        [1/2] /gpfs/tp/gwf/gwf_cmt/cmarsh/spack/stage/spack-stage-llvm-17.0.6-siym653rezmksycrvlez7jah5tk72os3/spack-build-siym653/bin/clang --target=x86_64-unknown-linux-gnu   --gcc-install-dir=/cvmfs/soft.computecanada.ca/gentoo/2023/x86-64-v3/usr/lib/gcc/x86_64-pc-linux-gnu/12 -fPIC -fno-semantic-interposition -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -Wimplicit-fallthrough -Wcovered-switch-default -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -Wall -Wcast-qual -Wformat-pedantic -Wimplicit-fallthrough -Wsign-compare -Wno-enum-constexpr-conversion -Wno-extra -Wno-pedantic -DCHECK_FUNCTION_EXISTS=hwloc_topology_init  --start-no-unused-arguments --unwindlib=none --end-no-unused-arguments -nostdlib++ -nostdinc++ -MD -MT CMakeFiles/cmTC_48e41.dir/CheckFunctionExists.c.o -MF CMakeFiles/cmTC_48e41.dir/CheckFunctionExists.c.o.d -o CMakeFiles/cmTC_48e41.dir/CheckFunctionExists.c.o -c /gpfs/tp/gwf/gwf_cmt/cmarsh/spack/stage/spack-stage-llvm-17.0.6-siym653rezmksycrvlez7jah5tk72os3/spack-build-siym653/runtimes/runtimes-bins/CMakeFiles/CMakeScratch/TryCompile-XydOOq/CheckFunctionExists.c
        [2/2] : && /gpfs/tp/gwf/gwf_cmt/cmarsh/spack/stage/spack-stage-llvm-17.0.6-siym653rezmksycrvlez7jah5tk72os3/spack-build-siym653/bin/clang --target=x86_64-unknown-linux-gnu --gcc-install-dir=/cvmfs/soft.computecanada.ca/gentoo/2023/x86-64-v3/usr/lib/gcc/x86_64-pc-linux-gnu/12 -fPIC -fno-semantic-interposition -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -Wimplicit-fallthrough -Wcovered-switch-default -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -Wall -Wcast-qual -Wformat-pedantic -Wimplicit-fallthrough -Wsign-compare -Wno-enum-constexpr-conversion -Wno-extra -Wno-pedantic -DCHECK_FUNCTION_EXISTS=hwloc_topology_init  --start-no-unused-arguments --unwindlib=none --end-no-unused-arguments -nostdlib++ -nostdinc++  CMakeFiles/cmTC_48e41.dir/CheckFunctionExists.c.o -o cmTC_48e41 -L/globalhome/cbm038/HPC/project/spack/install/linux-rhel9-skylake_avx512/gcc-12.3.1/hwloc-2.9.3-rpi4timtsf6t32tbrgvhkxcwfqxnywhv/lib -Wl,-rpath,/globalhome/cbm038/HPC/project/spack/install/linux-rhel9-skylake_avx512/gcc-12.3.1/hwloc-2.9.3-rpi4timtsf6t32tbrgvhkxcwfqxnywhv/lib  /globalhome/cbm038/HPC/project/spack/install/linux-rhel9-skylake_avx512/gcc-12.3.1/hwloc-2.9.3-rpi4timtsf6t32tbrgvhkxcwfqxnywhv/lib/libhwloc.so && :
        FAILED: cmTC_48e41 
        : && /gpfs/tp/gwf/gwf_cmt/cmarsh/spack/stage/spack-stage-llvm-17.0.6-siym653rezmksycrvlez7jah5tk72os3/spack-build-siym653/bin/clang --target=x86_64-unknown-linux-gnu --gcc-install-dir=/cvmfs/soft.computecanada.ca/gentoo/2023/x86-64-v3/usr/lib/gcc/x86_64-pc-linux-gnu/12 -fPIC -fno-semantic-interposition -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -Wimplicit-fallthrough -Wcovered-switch-default -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -Wall -Wcast-qual -Wformat-pedantic -Wimplicit-fallthrough -Wsign-compare -Wno-enum-constexpr-conversion -Wno-extra -Wno-pedantic -DCHECK_FUNCTION_EXISTS=hwloc_topology_init  --start-no-unused-arguments --unwindlib=none --end-no-unused-arguments -nostdlib++ -nostdinc++  CMakeFiles/cmTC_48e41.dir/CheckFunctionExists.c.o -o cmTC_48e41 -L/globalhome/cbm038/HPC/project/spack/install/linux-rhel9-skylake_avx512/gcc-12.3.1/hwloc-2.9.3-rpi4timtsf6t32tbrgvhkxcwfqxnywhv/lib -Wl,-rpath,/globalhome/cbm038/HPC/project/spack/install/linux-rhel9-skylake_avx512/gcc-12.3.1/hwloc-2.9.3-rpi4timtsf6t32tbrgvhkxcwfqxnywhv/lib  /globalhome/cbm038/HPC/project/spack/install/linux-rhel9-skylake_avx512/gcc-12.3.1/hwloc-2.9.3-rpi4timtsf6t32tbrgvhkxcwfqxnywhv/lib/libhwloc.so && :
        /cvmfs/soft.computecanada.ca/gentoo/2023/x86-64-v3/usr/x86_64-pc-linux-gnu/binutils-bin/2.41/ld: /cvmfs/soft.computecanada.ca/gentoo/2023/x86-64-v3/usr/lib/gcc/x86_64-pc-linux-gnu/12/libgcc_s.so.1: undefined reference to `_dl_find_object@GLIBC_2.35'
        /cvmfs/soft.computecanada.ca/gentoo/2023/x86-64-v3/usr/x86_64-pc-linux-gnu/binutils-bin/2.41/ld: /cvmfs/soft.computecanada.ca/gentoo/2023/x86-64-v3/usr/lib/gcc/x86_64-pc-linux-gnu/12/libstdc++.so.6: undefined reference to `arc4random@GLIBC_2.36'
        clang: error: linker command failed with exit code 1 (use -v to see invocation)
        ninja: build stopped: subcommand failed.
        
      exitCode: 1
...
```

I've run into this with libcxx 18.1.8 on linux, where dl gets misdetected:
```
-- Looking for dladdr in dl
-- Looking for dladdr in dl - not found
-- Looking for pthread_once in pthread
-- Looking for pthread_once in pthread - not found
```
and then compilation fails on missing symbols for `dladdr`. Looking under the hood of that detection failure, it's likewise due to `-nostdlib++`:
```
    kind: "try_compile-v1"
    backtrace:
      - "$BUILD_PREFIX/share/cmake-3.30/Modules/CheckLibraryExists.cmake:69 (try_compile)"
      - "$SRC_DIR/libcxxabi/cmake/config-ix.cmake:109 (check_library_exists)"
      - "$SRC_DIR/libcxxabi/CMakeLists.txt:249 (include)"
    checks:
      - "Looking for dladdr in dl"
    directories:
      source: "$SRC_DIR/build/CMakeFiles/CMakeScratch/TryCompile-dRIlTz"
      binary: "$SRC_DIR/build/CMakeFiles/CMakeScratch/TryCompile-dRIlTz"
    cmakeVariables:
      CMAKE_C_FLAGS: "-march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem $PREFIX/include -fdebug-prefix-map=$SRC_DIR=/usr/local/src/conda/libcxx_and_libcxxabi-18.1.8 -fdebug-prefix-map=$PREFIX=/usr/local/src/conda-prefix -I$PREFIX/include -I$BUILD_PREFIX/include -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -Wno-comment -fdiagnostics-color"
      CMAKE_C_FLAGS_DEBUG: "-g"
      CMAKE_EXE_LINKER_FLAGS: "-Wl,-O2 -Wl,--sort-common -Wl,--as-needed -Wl,-z,relro -Wl,-z,now -Wl,--disable-new-dtags -Wl,--gc-sections -Wl,--allow-shlib-undefined -Wl,-rpath,$PREFIX/lib -Wl,-rpath-link,$PREFIX/lib -L$PREFIX/lib -L$PREFIX/lib -Wl,-rpath,$PREFIX/lib -L$BUILD_PREFIX/lib -Wl,-rpath,$BUILD_PREFIX/lib"
      CMAKE_MODULE_PATH: "$SRC_DIR/libcxxabi/cmake;$SRC_DIR/libcxxabi/cmake/Modules;$SRC_DIR/libcxxabi/cmake;$SRC_DIR/libcxxabi/cmake/Modules;$SRC_DIR/libcxxabi/../runtimes/cmake/Modules;$SRC_DIR/libcxxabi/../cmake;$SRC_DIR/libcxxabi/../cmake/Modules;$SRC_DIR/runtimes/cmake;$SRC_DIR/runtimes/cmake/modules;$SRC_DIR/runtimes/../cmake;$SRC_DIR/runtimes/../cmake/Modules;$SRC_DIR/runtimes/../llvm/cmake;$SRC_DIR/runtimes/../llvm/cmake/modules"
    buildResult:
      variable: "LIBCXXABI_HAS_DL_LIB"
      cached: true
      stdout: |
        Change Dir: '$SRC_DIR/build/CMakeFiles/CMakeScratch/TryCompile-dRIlTz'
        
        Run Build Command(s): $BUILD_PREFIX/bin/ninja -v cmTC_c57b3
        [1/2] $BUILD_PREFIX/bin/x86_64-conda-linux-gnu-cc   -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem $PREFIX/include -fdebug-prefix-map=$SRC_DIR=/usr/local/src/conda/libcxx_and_libcxxabi-18.1.8 -fdebug-prefix-map=$PREFIX=/usr/local/src/conda-prefix -I$PREFIX/include -I$BUILD_PREFIX/include -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -Wno-comment -fdiagnostics-color -DCHECK_FUNCTION_EXISTS=dladdr  -nostdlib++ -nostdinc++ -nostdlib++ -o CMakeFiles/cmTC_c57b3.dir/CheckFunctionExists.c.o -c $SRC_DIR/build/CMakeFiles/CMakeScratch/TryCompile-dRIlTz/CheckFunctionExists.c
        FAILED: CMakeFiles/cmTC_c57b3.dir/CheckFunctionExists.c.o 
        $BUILD_PREFIX/bin/x86_64-conda-linux-gnu-cc   -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem $PREFIX/include -fdebug-prefix-map=$SRC_DIR=/usr/local/src/conda/libcxx_and_libcxxabi-18.1.8 -fdebug-prefix-map=$PREFIX=/usr/local/src/conda-prefix -I$PREFIX/include -I$BUILD_PREFIX/include -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -Wno-comment -fdiagnostics-color -DCHECK_FUNCTION_EXISTS=dladdr  -nostdlib++ -nostdinc++ -nostdlib++ -o CMakeFiles/cmTC_c57b3.dir/CheckFunctionExists.c.o -c $SRC_DIR/build/CMakeFiles/CMakeScratch/TryCompile-dRIlTz/CheckFunctionExists.c
        x86_64-conda-linux-gnu-cc: error: unrecognized command-line option '-nostdlib++'
        x86_64-conda-linux-gnu-cc: error: unrecognized command-line option '-nostdlib++'
        ninja: build stopped: subcommand failed.
        
      exitCode: 1
```

--- a/runtimes/CMakeLists.txt
+++ b/runtimes/CMakeLists.txt
@@ -160,13 +160,26 @@ endif()
 # Check for -nostdlib++ first; if there's no C++ standard library yet,
 # all check_cxx_compiler_flag commands will fail until we add -nostdlib++
 # (or -nodefaultlibs).
-llvm_check_compiler_linker_flag(CXX "-nostdlib++" CXX_SUPPORTS_NOSTDLIBXX_FLAG)
-if (CXX_SUPPORTS_NOSTDLIBXX_FLAG)
-  set(CMAKE_REQUIRED_FLAGS "${CMAKE_REQUIRED_FLAGS} -nostdlib++")
-endif()
-check_cxx_compiler_flag(-nostdinc++ CXX_SUPPORTS_NOSTDINCXX_FLAG)
-if (CXX_SUPPORTS_NOSTDINCXX_FLAG)
-  set(CMAKE_REQUIRED_FLAGS "${CMAKE_REQUIRED_FLAGS} -nostdinc++")
+#
+# CMAKE_REQUIRED_FLAGS is used both for C and C++ compilation. This breaks
+# compiling with GCC (and we should not need to force -nostd* for GCC anyway),
+# so apply it only with Clang. However, since there are cases when appending
+# the flag actually breaks the build, still perform the checks rather than
+# appending it unconditionally.
+#
+# TODO: find a better solution. See the discussion on:
+# https://github.com/llvm/llvm-project/issues/90332
+# https://github.com/llvm/llvm-project/pull/108357
+
+if ("${CMAKE_CXX_COMPILER_ID}" MATCHES "Clang")
+  llvm_check_compiler_linker_flag(CXX "-nostdlib++" CXX_SUPPORTS_NOSTDLIBXX_FLAG)
+  if (CXX_SUPPORTS_NOSTDLIBXX_FLAG)
+    set(CMAKE_REQUIRED_FLAGS "${CMAKE_REQUIRED_FLAGS} -nostdlib++")
+  endif()
+  check_cxx_compiler_flag(-nostdinc++ CXX_SUPPORTS_NOSTDINCXX_FLAG)
+  if (CXX_SUPPORTS_NOSTDINCXX_FLAG)
+    set(CMAKE_REQUIRED_FLAGS "${CMAKE_REQUIRED_FLAGS} -nostdinc++")
+  endif()
 endif()
 
 # Avoid checking whether the compiler is working.

